rust/src/server.rs

```diff
@@ -0,0 +1,190 @@
+    *response.status_mut() = http_resp.status();
+    // tracing::info!(
+    //     "Sent request to {} with Content-Encoding {}, got {}",
+    //     uri_string,
+    //     get_header(http_resp.headers(), "content-encoding", "unknown"),
+    //     http_resp.status()
+    // );
+    Ok(response)
+}
+
+fn get_header<'a>(headers: &'a HeaderMap, key: &'a str, default_value: &'a str) -> &'a str {
+    headers
+        .get(key)
+        .map_or(default_value, |v| v.to_str().unwrap())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures_channel::oneshot;
+    use hyper::{Client, Method, Request};
+    use mockito::{mock, server_address};
+    use std::{
+        net::{SocketAddr, TcpListener as StdTcpListener},
+        thread,
+    };
+    use tokio::net::TcpListener;
+
+    #[tokio::test]
+    async fn test_proxy_handle() {
+        let mock = mock("POST", "/some/test/path")
+            .match_body("{expected payload}")
+            .match_header("content-type", "application/test")
+            .match_header("DD-API-KEY", "SECRET API KEY")
+            .match_header("user-agent", "unit-test")
+            .with_body("{expected response}")
+            .with_status(418)
+            .expect(1)
+            .create();
+        let server = TestServer::serve(server_address());
+        let uri_string = format!("http://{}/some/test/path", server.addr);
+        let uri = uri_string
+            .parse::<hyper::Uri>()
+            .expect("server addr should parse");
+        let client = Client::new();
+        let req = Request::builder()
+            .method(Method::POST)
+            .header("content-type", "application/test")
+            .header("DD-API-KEY", "SECRET API KEY")
+            .header("user-agent", "unit-test")
+            .uri(uri)
+            .body(Body::from("{expected payload}"))
+            .expect("request builder");
+        let resp = client.request(req).await.unwrap();
+        assert_eq!(resp.status(), 418);
+        mock.assert();
+    }
+
+    pub fn runtime() -> tokio::runtime::Runtime {
+        tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .expect("new rt")
+    }
+
+    fn tcp_bind(addr: &SocketAddr) -> ::tokio::io::Result<TcpListener> {
+        let std_listener = StdTcpListener::bind(addr).unwrap();
+        std_listener.set_nonblocking(true).unwrap();
+        TcpListener::from_std(std_listener)
+    }
+
+    struct TestServer {
+        shutdown_signal: Option<oneshot::Sender<()>>,
+        thread: Option<thread::JoinHandle<()>>,
+        addr: SocketAddr,
+    }
+
+    impl TestServer {
+        fn serve(proxy_addr: SocketAddr) -> TestServer {
+            let (shutdown_tx, shutdown_rx) = oneshot::channel();
+            let listener = tcp_bind(&"127.0.0.1:0".parse().unwrap()).unwrap();
+            let addr = listener.local_addr().unwrap();
+            let thread = thread::Builder::new()
```

Comment:
Hmm, threads in an async context?  Could this be done with a tokio task?  Would that make the shutdown situation easier to handle?